### PR TITLE
EncryptingEntity and EncryptedPartEntity should not close writeTo(OutputStream)

### PR DESCRIPTION
**Pull request by tjcelaya**
This PR adds test cases for `EncryptingEntity` and an integration test for `EncryptingPartEntity`. 

Apologies for the large segment of changes in `EncryptingEntityTest.java` I noticed the last test case I previously added (`canBeWrittenIdempotently`) was after the utility methods so I moved it to the right place before adding `doesNotCloseSuppliedOutputStreamWhenFailureOccurs` and `doesNotCloseSuppliedOutputStreamWhenWrittenSuccessfully`.

These entities should be writing content to the provided `OutputStream` but *not* closing it, otherwise requests using chunked encoding (those backed by an `InputStream`) complete successfully even if they were intended to fail. This results in intermittent test failures during the MPU commit operation where the intentionally-failing payloads in `EncryptedServerSideMultipartManagerIT#canRetryUploadPart` can "win out" over the full part content. This is visible in the following directory listing of an upload directory for a commit which failed on 409 Conflict as described in #332:

```
mls -l ~~/uploads/2/20c5f2f5-efe3-cc72-fe73-e89a7a4bc123
-rwxr-xr-x 1 java.manta.ci          1024 Sep 11 15:30 0
-rwxr-xr-x 1 java.manta.ci       5242880 Sep 11 15:30 1
-rwxr-xr-x 1 java.manta.ci           512 Sep 11 15:30 2
```

The `1024` byte part 0 would not have been created if `EncryptingPartEntity` had left the `OutputStream` intact (so that HttpClient could terminate the request correctly).

### CODE REVIEW

**Comment by tjcelaya** [comment]
(This is the client-side resolution for MANTA-3423, server-side investigation still pending)

**Comment by cburroughs** [comment]
If `HttpEntity.writeTo` isn't supposed to close (the [javadocs](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpEntity.html#writeTo-java.io.OutputStream-) are sadly not clarifying), do we know what will do the close?

Do we know why (aka what code) looked at the outputstream and said "this was closed so everything is good and complete) which caused the original issue?

**Comment by tjcelaya** [comment]
The correct place for the provided `OutputStream` to be closed is in [DefaultBHttpClientConnection#sendRequestEntity](https://github.com/apache/httpcomponents-core/blob/4.4.x/httpcore/src/main/java/org/apache/http/impl/DefaultBHttpClientConnection.java#L147) which builds the `OutputStream`, calls `HttpEntity#writeTo` with it and then calls `close` on the stream. In the face of network or input errors, an exception would be thrown which prevents the regular call to `close`, but the catch causes in `MainClientExec#execute` are supposed to [abort the connection](https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient/src/main/java/org/apache/http/impl/execchain/MainClientExec.java#L344) before they rethrow. In the case of `ChunkedOutputStream`, calling `close` when an error occurs is *plain wrong* because it insinuates that the request terminated successfully.

File: java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerIT.java [hunk 1]
```diff
@@ -576,4 +578,27 @@ public final void canRetryUploadPart() throws IOException {
...
+        EncryptedMultipartUpload<ServerSideMultipartUpload> upload = multipart.initiateUpload(path);
+
+        Assert.assertThrows(IOException.class, () -> {
+            // partial read of content1
+            multipart.uploadPart(upload, 1, new FailingInputStream(new ByteArrayInputStream(content), 1024));
+        });
+
+        // listing the directory immediately after MPU creation is likely to return no results
+        // this is being reported upstream
+        TimeUnit.SECONDS.sleep(1);
```
Comment (dekobon) [inline java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerIT.java]: Shouldn't you do a loop around the sleep in case your sleep wait value is too small?

Comment (tjcelaya) [reply java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerIT.java]: This is a really weird context to try to explain..

Based on the fact that you need to see a test fail before it passes I was trying to design the test so it would fail without the proposed change (during which I temporarily reverted the fix). While running the test case I discovered that the uploaded part does not appear in the MPU part directory immediately after the PUT request completes, so the test case was passing both *with* and *without* the change. 

The [RFD](https://github.com/joyent/rfd/blob/master/rfd/0065/README.md#2-upload-part) seems to imply that there might be a race condition between upload-part and MPU state. I couldn't find anything that might imply directory listings would be affected though.

In summary, we don't really have anything to check in a loop since the expectation is that the part never appears. This test is starting to seem a bit contrived, I'll need to rethink whether it's worth keeping. Perhaps it would make more sense to create a unit test for `EncryptingPartEntity` like the test cases created for `EncryptingEntity`?

File: java-manta-client-unshaded/src/test/java/com/joyent/manta/client/multipart/EncryptingPartEntityTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package com.joyent.manta.client.multipart;
```
Comment (cburroughs) [inline java-manta-client-unshaded/src/test/java/com/joyent/manta/client/multipart/EncryptingPartEntityTest.java]: ^^ License headers

File: java-manta-client-unshaded/src/test/java/com/joyent/manta/client/multipart/EncryptingPartEntityTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                state.getMultipartStream(),
+                new InputStreamEntity(new BrokenInputStream(new IOException("bad input"))),
+                CALLBACK_NOOP);
+
+        final OutputStream output = Mockito.mock(OutputStream.class);
+        Assert.assertThrows(IOException.class, () -> encryptingPartEntity.writeTo(output));
+        Mockito.verify(output, Mockito.never()).close();
+    }
+
+}
```
Comment (cburroughs) [inline java-manta-client-unshaded/src/test/java/com/joyent/manta/client/multipart/EncryptingPartEntityTest.java]: nit: newline

File: java-manta-client-unshaded/src/test/java/com/joyent/manta/client/crypto/EncryptingEntityTest.java [hunk 4]
```diff
@@ -116,6 +120,89 @@ public void canSurviveNetworkFailuresInAesCbc() throws Exception {
...
+
+    private void canBeWrittenIdempotently(final SupportedCipherDetails cipherDetails) throws Exception {
+        final SecretKey secretKey = SecretKeyUtils.generate(cipherDetails);
+        final String content = RandomStringUtils.randomAlphanumeric(RandomUtils.nextInt(500, 1500));
+        final ExposedStringEntity contentEntity = new ExposedStringEntity(
+                content,
+                StandardCharsets.UTF_8);
+
+        final ByteArrayOutputStream referenceEncrypted = new ByteArrayOutputStream(content.length());
+        {
```
Comment (cburroughs) [inline java-manta-client-unshaded/src/test/java/com/joyent/manta/client/crypto/EncryptingEntityTest.java]: I'm not following what this bracket signifies.

Comment (tjcelaya) [reply java-manta-client-unshaded/src/test/java/com/joyent/manta/client/crypto/EncryptingEntityTest.java]: This test was introduced with #296 and moved in this PR when I noticed it was below the test utility methods. The reason for the bracket was to isolate variable scopes, by the time I wrote this test I had been bitten by variable naming that made some tests seem to pass when they were supposed to fail and vice-versa. The braces made mistakenly-reused variables from the first round of encryption into compiler errors so I could be sure I wasn't mixing objects between encryption operations.
